// @nomadreservations/ngx-stripe/esm2015/projects/ngx-stripe/lib/ngx-stripe.module.js
var NgxStripeModule_1;
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { STRIPE_OPTIONS, STRIPE_PUBLISHABLE_KEY } from './interfaces/stripe';
import { PaymentRequestComponent } from './payment-request.component';
import { LazyStripeAPILoader } from './services/api-loader.service';
import { DocumentRef } from './services/document-ref.service';
import { PlatformService } from './services/platform.service';
import { StripeService } from './services/stripe.service';
import { WindowRef } from './services/window-ref.service';
import { StripeCardComponent } from './stripe-card.component';
let NgxStripeModule = NgxStripeModule_1 = class NgxStripeModule {
    static forRoot(publishableKey, options) {
        return {
            ngModule: NgxStripeModule_1,
            providers: [
                LazyStripeAPILoader,
                StripeService,
                PlatformService,
                WindowRef,
                DocumentRef,
                {
                    provide: STRIPE_PUBLISHABLE_KEY,
                    useValue: publishableKey
                },
                {
                    provide: STRIPE_OPTIONS,
                    useValue: options
                }
            ]
        };
    }
};
NgxStripeModule = NgxStripeModule_1 = tslib_1.__decorate([
    NgModule({
        imports: [CommonModule],
        declarations: [StripeCardComponent, PaymentRequestComponent],
        exports: [StripeCardComponent, PaymentRequestComponent]
    })
], NgxStripeModule);
export { NgxStripeModule };
